### PR TITLE
refactor: make instance started separately

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

to support further integrated into other binaries

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/cmd/src/cli.rs [hunk 1]
```diff
@@ -30,27 +38,31 @@ use upgrade::UpgradeCommand;
...
-        }
+    fn new(tool: Arc<dyn Tool>) -> Self {
+        Self { tool }
     }
+}
 
-    pub async fn stop(&self) -> Result<()> {
+#[async_trait]
+impl App for Instance {
+    async fn start(&self) -> Result<()> {
```
Comment (evenyag) [inline src/cmd/src/cli.rs:61]: Can `start` still takes `&mut self`?

**Comment by MichaelScofield** [comment]
@WenyXu @evenyag PTAL

File: src/cmd/src/cli/bench.rs [hunk 2]
```diff
@@ -69,7 +69,7 @@ impl BenchTableMetadataCommand {
             table_metadata_manager,
             count: self.count,
         };
-        Ok(Instance::Tool(Box::new(tool)))
+        Ok(Instance::new(Arc::new(tool)))
```
Comment (WenyXu) [inline src/cmd/src/cli/bench.rs:72]: Should be `Box`?

File: src/cmd/src/cli/export.rs [hunk 3]
```diff
@@ -105,7 +105,7 @@ impl ExportCommand {
             }));
         }
 
-        Ok(Instance::Tool(Box::new(Export {
+        Ok(Instance::new(Arc::new(Export {
```
Comment (WenyXu) [inline src/cmd/src/cli/export.rs:108]: ditto

File: src/cmd/src/cli/upgrade.rs [hunk 4]
```diff
@@ -76,7 +76,7 @@ impl UpgradeCommand {
             skip_schema_keys: self.skip_schema_keys,
             skip_table_route_keys: self.skip_table_route_keys,
         };
-        Ok(Instance::Tool(Box::new(tool)))
+        Ok(Instance::new(Arc::new(tool)))
```
Comment (WenyXu) [inline src/cmd/src/cli/upgrade.rs:79]: ditto

File: src/cmd/src/standalone.rs [hunk 5]
```diff
@@ -382,10 +393,11 @@ impl StartCommand {
         })
     }
 
-    async fn create_ddl_task_executor(
+    pub async fn create_ddl_task_executor(
```
Comment (WenyXu) [inline src/cmd/src/standalone.rs:396]: Do we need to make it as `public`?
```suggestion
    async fn create_ddl_task_executor(
```

File: src/cmd/src/standalone.rs [hunk 6]
```diff
@@ -230,14 +237,14 @@ struct StartCommand {
...
-    env_prefix: String,
+    pub env_prefix: String,
     /// The working home directory of this standalone instance.
     #[clap(long)]
     data_home: Option<String>,
 }
 
 impl StartCommand {
-    fn load_options(&self, top_level_options: TopLevelOptions) -> Result<Options> {
+    pub fn load_options(&self, cli_options: &CliOptions) -> Result<Options> {
```
Comment (WenyXu) [inline src/cmd/src/standalone.rs:247]: ditto
```suggestion
    fn load_options(&self, cli_options: &CliOptions) -> Result<Options> {
```

File: src/cmd/src/standalone.rs [hunk 7]
```diff
@@ -324,7 +331,7 @@ impl StartCommand {
     #[allow(unreachable_code)]
     #[allow(unused_variables)]
     #[allow(clippy::diverging_sub_expression)]
-    async fn build(self, opts: MixOptions) -> Result<Instance> {
+    pub async fn build(self, opts: MixOptions) -> Result<Instance> {
```
Comment (WenyXu) [inline src/cmd/src/standalone.rs:334]: Ditto. Do we need to make it as public? Or Let's add some comment.

**Review by WenyXu — APPROVED** [review]
Rest LGTM

File: src/cmd/src/standalone.rs (hunk 5 above)
Comment (MichaelScofield) [reply src/cmd/src/standalone.rs:396]: Yes, it's used in another project.

**Review by evenyag — APPROVED** [review]

**Pull request merged**